Wobble/Scheduling/TaskHandler.cs

```diff
@@ -0,0 +1,119 @@
+
+        /// <summary>
+        ///     Determined by if current task is completed.
+        /// </summary>
+        public bool IsCompleted { get; private set; }
+
+        /// <summary>
+        ///     Determined by if the current task is running.
+        /// </summary>
+        public bool IsRunning { get; private set; }
+
+        /// <summary>
+        ///     Determined by if the current task is cancelled.
+        /// </summary>
+        public bool IsCancelled { get; private set; }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="function"></param>
+        public TaskHandler(Func<T, TResult> function) => Function = function;
+
+        /// <summary>
+        ///     This method will cancel the previous task and run a new task.
+        /// </summary>
+        /// <param name="input"></param>
+        /// <param name="delay"></param>
+        public void Run(T input, int delay = 0)
+        {
+            // We want to automatically cancel the previous task and dispose the other
+            // cancellation token source before running a new task.
+            Source.Cancel();
+            Source.Dispose();
+            Source = new CancellationTokenSource();
+            IsCompleted = false;
+            IsCancelled = false;
+            IsRunning = true;
+
+            Task.Run(async () =>
+            {
+                try
+                {
+                    var token = Source.Token;
+                    await Task.Delay(delay, token);
+
+                    var result = Function(input);
+                    token.ThrowIfCancellationRequested();
+                    IsCompleted = true;
+                    IsRunning = false;
+                    OnCompleted?.Invoke(typeof(TaskHandler<T, TResult>), new TaskCompleteEventArgs<T, TResult>(input, result));
+                }
+                catch (OperationCanceledException e)
+                {
+                    OnSourceCancelled(input);
+                }
+                catch (Exception e)
+                {
+                    Logger.Error(e, LogType.Runtime);
+                }
+            });
+        }
+
+        /// <summary>
+        ///     This method is called when a task is cancelled or if user wants to cancel existing task.
+        /// </summary>
+        /// <param name="input"></param>
+        private void OnSourceCancelled(T input)
+        {
+            IsCancelled = true;
+            IsRunning = false;
+            OnCancelled?.Invoke(typeof(TaskHandler<T, TResult>), new TaskCancelledEventArgs<T>(input));
+        }
+
+        /// <summary>
+        ///     Cancel the current existing task.
+        /// </summary>
+        public void Cancel() => Source.Cancel();
+
+        /// <summary>
+        ///     Should be called when Task Handler will no longer be used.
+        /// </summary>
+        public void Dispose()
+        {
+            Source.Cancel();
```

Comment:
Same issue with synchronization: it's probably desirable here to still have the cancelled thread run the handlers and not just set them to `null`.

Reply:
Do you mean to not cancel the source? Only thing set to null are the event handlers

Reply:
This comment is placed a little off. I mean that when someone `Dispose`s a running task it is cancelled but the `OnCancelled` callbacks aren't run because they are set to `null` here. I'm suggesting to not set them to `null` so they are still run.

Reply:
This also suffers from the same issue as the other comment and needs `Source.Cancel()` in a `lock (this)`.